### PR TITLE
Implement issue #297 governance V1 hooks

### LINKED ISSUE #297: Bot Society Governance V1: safety boundary, permission, and accountability for interbot DM
## Summary

MEP interbot DM has now proven that bots can `talk` and can also enter a real `talk + work` path. However, as this capability becomes more useful, the system needs a clearer governance model for privacy, sensitive context sharing, and action safety.

Today, the main gap is not only transport or execution routing. The larger gap is that bot-to-bot collaboration still lacks a shared control layer similar to how human society works:
- do not over-share by default
- ask before disclosing sensitive information
- keep an audit trail
- establish trust step by step rather than assuming full openness

This issue proposes a staged governance roadmap for MEP bot society, starting with a minimal V1 focused on control and safety.

## Problem

Current interbot behavior is powerful but under-governed:

- `talk` is lower-risk than `talk + work`, but it can still disclose sensitive information.
- `talk + work` can now trigger real execution paths, but sensitive workspace facts and local context may need to cross bot boundaries.
- human-chat paths often have richer local session context than interbot DM paths, which creates pressure to share more context over MEP.
- without clear rules, the system risks two bad extremes:
  - too closed: bots are safe but not useful
  - too open: bots are useful but leak context or overreach

The system needs a shared governance model before adding richer trust or collaboration features.

### V1 Goals
Build the minimum control layer for interbot collaboration:

- define a minimal `Safety Boundary`
- require explicit `Permission` for sensitive disclosure or risky actions
- add `Accountability` for what was shared, approved, and executed

### V2 Goals
Improve the social behavior of bots:

- add `Politeness` defaults such as minimal disclosure and ask-first behavior
- begin a step-by-step `Trust` model based on stable identity and history

### V3 Goals
Improve collaboration quality:

- add `Competence` signals for routing tasks to the right bots
- combine trust and competence for higher-quality multi-bot coordination

## Non-Goals

This issue does **not** aim to implement all governance layers at once.

Not in scope for V1:
- full trust ladder / reputation engine
- complete bot competence scoring system
- complex global policy engine
- automatic free-form sharing of full session memory or full chat history
- replacing human approval for high-risk disclosures

### 1. Safety Boundary
Define what is:
- `safe`
- `approval_required`
- `forbidden`

Examples:
- safe:
  - task goal
  - repo URL
  - relative file paths
  - non-sensitive blocker summary
- approval required:
  - local absolute paths
  - session memory excerpts
  - raw logs
  - runtime environment details
- forbidden:
  - secrets
  - private keys
  - raw tokens
  - unsafe destructive actions without explicit approval

### 2. Permission
When a bot needs sensitive information or a risky action, it should not proceed automatically.

Instead it should emit a structured approval request, such as:
- what is needed
- why it is needed
- scope of disclosure
- risk level
- which task it applies to

Human approval should be explicit and auditable.

### 3. Accountability
The system should record:
- who requested disclosure
- who approved or denied it
- what category of information was shared
- what action was executed
- for which task / context
- the stated reason

This creates reviewability and supports future trust building.

### 4. Politeness
Future layer:
- minimal disclosure by default
- ask before reading or sharing more
- avoid oversharing unrelated context
- avoid unnecessary interruption when local discovery is sufficient

### 5. Trust
Future layer:
- stable identity
- behavior history
- task-specific trust
- progressive disclosure based on proven reliability

### 6. Competence
Future layer:
- route tasks based on demonstrated capability
- distinguish bots trusted for review, execution, deploy, audit, etc.

## Why Start With V1

The first priority is not social sophistication; it is control.

Before trust ladders or advanced routing, the system needs:
- hard boundaries
- explicit approval flow
- audit logs

Without these, richer bot-to-bot collaboration becomes unsafe and difficult to govern.

### A. Minimal classification model
Add a minimal shared classification for interbot context/action handling:
- `safe`
- `approval_required`
- `forbidden`

### B. Approval request / decision flow
Add a structured way for bots to request approval before:
- sharing sensitive context
- exposing local runtime facts
- performing risky execution steps

### C. Audit trail
Record:
- disclosure requests
- approval decisions
- sensitive action execution
- rationale and task linkage

### D. Fail-closed behavior
If content or action cannot be safely classified:
- do not proceed automatically
- require approval or return a blocker

### Scenario 1: Bot needs a local absolute repo path
- Bot tries local discovery first
- If unresolved, it emits an approval request
- Human may approve disclosure of the path only
- Disclosure is logged

### Scenario 2: Bot wants to send session memory to another bot
- Default behavior should be summary-first
- Raw memory content should require approval
- If uncertain whether it contains sensitive content, fail closed

### Scenario 3: Execution task wants to edit a workspace
- Normal task instructions may be safe
- destructive or environment-sensitive actions require explicit approval
- all approvals and execution results are logged

## Acceptance Criteria For V1

- interbot disclosure rules distinguish at least:
  - `safe`
  - `approval_required`
  - `forbidden`
- bots can emit a structured approval request for sensitive disclosure/actions
- approval decisions can be recorded and linked to a task
- sensitive actions and disclosures are auditable
- uncertain cases fail closed rather than defaulting to disclosure
- V1 does not require a full trust/reputation system

## Open Questions

- which runtime facts should always require approval versus be auto-redacted?
- should approval scope be one-shot, task-scoped, or time-scoped?
- how should redaction interact with approval flow?
- should talk-only and talk+work share the exact same governance envelope, or should work actions have stronger defaults?

## Suggested Follow-Up

After agreement on this issue:
- implement a narrow V1 PR for:
  - safety boundary
  - permission flow
  - accountability trail
- defer trust ladder, competence routing, and advanced bot-society behavior to later issues/PRs

### PR DESCRIPTION
**Pull request by WUAIBING**
## Summary
- implement the first V1 slice for #297 on interbot DM / talk+work paths
- add client-side governance metadata helpers for `safe`, `approval_required`, and `forbidden`
- add hub-side fail-closed validation and governance audit logging for interbot payloads
- cover the new behavior with client and hub API tests

## What Changed
- `clients/shared/mep_client.py`
  - add governance metadata builders/extractors
  - allow interbot and execution DMs to carry governance metadata
  - attach default `approval_required + pending` governance metadata to human approval requests
- `hub/main.py`
  - validate interbot governance metadata when interbot spec validation is enabled
  - reject `forbidden` governance payloads
  - require `approval.status=approved` for non-human-approval `approval_required` payloads
  - emit governance and human-approval audit events on submit/result paths
- tests
  - add client tests for governance metadata round-trip
  - add hub tests for forbidden fail-closed behavior and approval-request audit logging

## Validation
- `python -m pytest tests/test_shared_mep_client.py tests/test_hub_api.py tests/test_execution_bridge.py -q`

Closes #297

### CODE REVIEW

**Comment by WUAIBING** [comment]
@Hub Sentinel review
@Elsaws Bot review

**Review by Hub-Sentinel — COMMENTED** [review]
## Review Summary

The PR implements governance V1 hooks for interbot DM/talk+work paths, adding client-side metadata helpers and hub-side validation. The changes appear correct and well-tested, with no high-impact issues found after verification.

Observation: In `clients/shared/mep_client.py`, the `build_governance_metadata` function accepts `classification` and `approval_status` without validating them against the allowed sets. While this could lead to silent misclassification, the hub-side validation in `hub/main.py` catches invalid values, and the tests cover the expected paths.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: correctness/regression around the changed behavior, missing or misaligned test coverage for the changed behavior, API contract, validation, and fallback-path drift

Checks performed: Verified that `build_governance_metadata` in `clients/shared/mep_client.py` constructs the metadata dict correctly witho, Checked that hub-side interbot governance validation in `hub/main.py` does not break existing interbot DMs without gover, Reviewed test coverage in `tests/test_hub_api.py` and `tests/test_shared_mep_client.py`; the tests cover `forbidden` rej, Confirmed that the `approval_status` validation is not enforced client-side, but hub-side audit logging would catch inva

Changed identifiers verified: `build_governance_metadata`, `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`

Why no finding: The candidate risks were verified against the actual diff and tests. The client-side lack of validation for `classification` and `approval_status` is mitigated by hub-side validation and feature gating. The missing test for `safe` classification is low-risk because the hub defaults to allowing it.

**Review by wuyanbingep-a11y — COMMENTED** [review]
## Review Summary

PR introduces interbot governance metadata and hub validation with a notable maintenance risk from duplicated constant definitions across client and hub modules.

Observation: Client (mep_client.py) and hub (hub/main.py) both define identical constant sets for governance classifications and approval statuses, with the hub additionally defining INTERBOT_REVIEW_VERDICTS. This creates a maintenance risk where updates to allowed values require manual synchronization across both modules.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: constant drift between client/hub modules, validation edge cases in hub governance checks, test coverage completeness for new governance paths

Checks performed: Compared constant definitions between client and hub, Verified test existence for forbidden governance rejection, Reviewed PR description for validation approach and test coverage claims

Changed identifiers verified: `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`, `INTERBOT_GOVERNANCE_APPROVAL_STATUSES`

Why no finding: No concrete correctness, regression, or missing-validation issue was supported after checking `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS` and `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`.

**Comment by WUAIBING** [comment]
@Hub Sentinel review\n@Elsaws Bot review\n(post-PR299 deploy rerun)

**Review by Hub-Sentinel — COMMENTED** [review]
## Review Findings

The PR implements governance V1 hooks for interbot DM/talk+work paths with client-side metadata helpers and hub-side validation. The changes are well-structured and include test coverage. No high-severity correctness or regression issues were found in the reviewed diff.

Observation: In `hub/main.py`, the validation logic for interbot governance metadata uses `governance.get('classification')` which returns `None` for missing keys, but the subsequent check `if classification not in INTERBOT_GOVERNANCE_CLASSIFICATIONS` will raise a `TypeError` because `None` is not hashable for set membership. This is a concrete edge case that could cause a crash on malformed payloads.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: Case sensitivity mismatch between client normalization and hub validation, Missing governance key handling in hub-side validation, Default governance metadata correctness for approval requests, Test coverage for invalid classification rejection

Checks performed: Verified that client-side `normalized_governance` lowercases classification while hub-side set is all lowercase, confirm, Checked hub validation code for `governance.get('classification')` and found potential `TypeError` when classification i, Reviewed test `test_interbot_governance_forbidden_rejects_direct_dm` to confirm it provides a governance key, Examined `build_governance_metadata` for input validation and found it accepts any string without checking against `GOVE

Changed identifiers verified: `GOVERNANCE_CLASSIFICATIONS`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`, `normalized_governance`, `build_governance_metadata`

1. **Potential TypeError when governance classification is missing or None.** (`hub/main.py`): The hub-side validation likely uses `governance.get('classification')` which returns `None` if the key is absent. The subsequent check `if classification not in INTERBOT_GOVERNANCE_CLASSIFICATIONS` will raise a `TypeError` because `None` is not hashable for set membership. This can crash the server on malformed interbot payloads.

File: hub/main.py [hunk 1]
```diff
@@ -897,6 +901,224 @@ def _interbot_require_string(data: dict, field_name: str) -> str:
     return value.strip()
 
 
+def _parse_interbot_message(payload_text: str) -> Optional[dict]:
+    try:
+        parsed = json.loads(payload_text)
+    except (TypeError, ValueError):
```
Comment (Hub-Sentinel) [inline hub/main.py:907]: **Potential TypeError when governance classification is missing or None.**

The hub-side validation likely uses `governance.get('classification')` which returns `None` if the key is absent. The subsequent check `if classification not in INTERBOT_GOVERNANCE_CLASSIFICATIONS` will raise a `TypeError` because `None` is not hashable for set membership. This can crash the server on malformed interbot payloads.

**Review by wuyanbingep-a11y — COMMENTED** [review]
## Review Summary

{ "summary": "The PR adds client-side governance metadata builders and hub-side validation for interbot payloads. The changes consistently implement the described governance V1 hooks with proper fail-closed validation and test coverage.", "observation": "The hub's INTERBOT_GOVERNANCE_CLASSIFICATIONS and INTERBOT_GOVERNANCE_APPROVAL_STATUSES constants in hub/main.py mirror the client's GOVERNANCE_CLASSIFICATIONS and GOVERNANCE_APPROVAL_STATUSES, ensuring consistent validation across components.",.

Risk areas checked: correctness/regression around the changed behavior, missing or misaligned test coverage for the changed behavior, API contract, validation, and fallback-path drift

Why no finding: No concrete correctness, regression, or missing-validation issue was directly supported by the supplied diff.

**Comment by WUAIBING** [comment]
@Hub Sentinel review this PR
@Elsaws Bot review this PR

**Review by Hub-Sentinel — APPROVED** [review]
## Review Summary

The PR introduces governance metadata support for interbot DM/talk+work paths with client-side helpers and hub-side validation. The changes are well-structured and covered by tests. No high-impact correctness or regression issues were found.

Observation: In `hub/main.py`, the validation logic for interbot governance metadata uses `INTERBOT_GOVERNANCE_CLASSIFICATIONS` directly without normalizing the incoming `classification` field. However, the client-side `build_governance_metadata` already normalizes via `normalized_classification`, so the hub-side validation is consistent as long as clients use the helper.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: correctness/regression around governance metadata normalization and validation, missing or misaligned test coverage for governance payload transmission, API contract and fallback-path drift for malformed governance metadata

Checks performed: Verified that `build_governance_metadata` normalizes classification via `normalized_classification` in `clients/shared/m, Checked that hub-side validation in `hub/main.py` uses `INTERBOT_GOVERNANCE_CLASSIFICATIONS` set for membership checks, Reviewed test `test_submit_dm_can_attach_governance_metadata` in `tests/test_shared_mep_client.py` to confirm it exercis, Examined test `test_interbot_governance_forbidden_rejects_direct_dm` in `tests/test_hub_api.py` for hub-side rejection l

Changed identifiers verified: `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `build_governance_metadata`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`.

**Comment by WUAIBING** [comment]
@Elsaws Bot review this PR

**Review by wuyanbingep-a11y — COMMENTED** [review]
## Review Summary

PR implements governance V1 metadata for interbot communications with consistent constant definitions between client and hub, validated by comprehensive tests covering forbidden/rejected paths and metadata attachment.

Observation: Both client and hub define identical governance classification and approval status constant sets, ensuring consistency across the API contract.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: API contract drift between client and hub constant definitions, Missing validation for governance metadata in interbot paths, Insufficient test coverage for rejection/forbidden scenarios

Checks performed: Verified GOVERNANCE_CLASSIFICATIONS/GOVERNANCE_APPROVAL_STATUSES constants match between client and hub, Confirmed hub validation logic rejects forbidden governance payloads, Reviewed test cases for both forbidden rejection and metadata attachment paths

Changed identifiers verified: `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`, `test_interbot_governance_forbidden_rejects_direct_dm`

Why no finding: The changed behavior appears safe because: 1) Client and hub governance constant sets (GOVERNANCE_CLASSIFICATIONS, GOVERNANCE_APPROVAL_STATUSES) are identical and synchronized, 2) Hub validation properly implements fail-closed rejection for forbidden payloads with explicit test coverage, 3) Client correctly attaches default governance metadata with test verification, and 4) All validation paths ar.

**Review by wuyanbingep-a11y — COMMENTED** [review]
## Review Summary

This PR adds V1 governance metadata helpers for client-side metadata building and hub-side validation for interbot payloads, with new test coverage for forbidden governance classification rejection and metadata attachment. Changes are properly scoped behind interbot spec validation checks.

Observation: The hub-side validation for INTERBOT_GOVERNANCE_CLASSIFICATIONS and INTERBOT_GOVERNANCE_APPROVAL_STATUSES appears to correctly enforce fail-closed semantics for interbot DMs when interbot spec validation is enabled, as evidenced by the test_interbot_governance_forbidden_rejects_direct_dm test case in tests/test_hub_api.py.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: fail-closed validation breaking older interbot clients, consistency between client-side builders and hub-side validation, completeness of governance field validation, test coverage edge cases

Checks performed: Examined hub/main.py constants for INTERBOT_GOVERNANCE_CLASSIFICATIONS, INTERBOT_GOVERNANCE_APPROVAL_STATUSES, INTERBOT_, Reviewed test_interbot_governance_forbidden_rejects_direct_dm in tests/test_hub_api.py to confirm governance metadata re, Verified that build_governance_metadata in clients/shared/mep_client.py uses GOVERNANCE_CLASSIFICATIONS and GOVERNANCE_A, Checked that PR description indicates validation is conditional on interbot spec validation being enabled

Changed identifiers verified: `INTERBOT_GOVERNANCE_CLASSIFICATIONS`, `INTERBOT_GOVERNANCE_APPROVAL_STATUSES`, `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`

Why no finding: The hub-side validation is conditional on interbot spec validation being enabled, which means existing interbot DMs without governance metadata won't be rejected unless the feature flag is active. The test coverage includes the core forbidden governance classification rejection case, and the client/hub validation sets appear consistent based on the provided diff excerpts.

**Review by wuyanbingep-a11y — COMMENTED** [review]
## Review Summary

PR implements governance V1 hooks for interbot DM and talk+work paths with client-side metadata helpers and hub-side validation. Changes add constants, parameters, and tests as described, with CI checks green. Review focused on validation correctness, constant drift, and test coverage but could not verify fail-closed behavior due to limited hub validation code visibility.

Observation: Both client (GOVERNANCE_CLASSIFICATIONS) and hub (INTERBOT_GOVERNANCE_CLASSIFICATIONS) define identical classification sets independently; this design choice may require careful synchronization if values change in the future.

Touched paths reviewed: `clients/shared/mep_client.py`, `hub/main.py`, `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Tests reviewed: `tests/test_hub_api.py`, `tests/test_shared_mep_client.py`

Risk areas checked: validation/fail-closed drift in hub governance checks, constant duplication between client and hub, test coverage for approval_status validation, default metadata consistency in build_governance_metadata

Checks performed: Examined patch excerpts for hub validation code (not fully visible), Verified constant sets match between client and hub definitions, Reviewed test coverage for forbidden classification rejection, Checked PR CI status and changed file list for completeness

Changed identifiers verified: `GOVERNANCE_CLASSIFICATIONS`, `GOVERNANCE_APPROVAL_STATUSES`, `INTERBOT_GOVERNANCE_CLASSIFICATIONS`, `INTERBOT_GOVERNANCE_APPROVAL_STATUSES`

Why no finding: The hub validation logic is not fully visible in provided excerpts, so fail-closed behavior cannot be verified. Constant duplication is a design choice, not a correctness risk. Tests for approval_status validation may exist but are not shown. Client default metadata behavior cannot be assessed without build_governance_metadata implementation.

**Review by wuyanbingep-a11y — APPROVED** [review]
## Review — Elsaws (deepseek/deepseek-v4-pro)

Read the full diff (565 lines, 4 files). Good governance V1 implementation. Clear, well-tested, fail-closed semantics.

### What works

- **Client-side** `build_governance_metadata` validates classification and approval_status against known sets, normalizes input, rejects invalid values with clear errors
- **Hub-side** `_extract_interbot_governance` validates with the same sets, rejects `forbidden` (403), enforces `approval_required` → `approved` except for human-approval intents where `pending` is correct
- **Audit logging** (`_log_interbot_governance_submit`/`_log_interbot_governance_result`) properly structured — HTTPException re-raised, other exceptions caught and logged without breaking task flow
- **Human approval requests** now auto-attach governance metadata (`approval_required`/`pending`) — good hygiene
- **Tests** cover forbidden rejection (403), human approval audit logging, metadata round-trip, and extraction

### What was flagged but is actually fine

- **Hub-Sentinel flagged potential `TypeError`** with `classification not in INTERBOT_GOVERNANCE_CLASSIFICATIONS` when classification is `None`.  This doesn't manifest — `_extract_interbot_governance` guards with `isinstance(classification, str)` before the set membership check. Missing key → `None` → `isinstance` returns `False` → clean HTTP 400.
- **Constant duplication** between client and hub — design smell, not a bug. Both sets are identical and tiny (3 values each). Consider extracting to a shared module in a follow-up, but not blocking.

### Minor nit (non-blocking)

`build_governance_metadata` doesn't prevent attaching an `approval` block to a `safe` classification. The hub-side validation ignores approval status for non-`approval_required` classifications, so no correctness impact — just a bit odd semantically. Fine for V1.

**Verdict: APPROVE.** Clean governance first-slice. Ship it.
